rust/src/lib/nm/profile_search.rs

```diff
@@ -0,0 +1,367 @@
+                        ret.push(nm_conn);
+                    }
+                }
+            }
+        }
+
+        ret.sort_unstable_by(|a, b| a.uuid().cmp(&b.uuid()));
+        ret.dedup();
+        ret
+    }
+
+    pub(crate) fn get_nm_ac(
+        &self,
+        iface: &BaseInterface,
+    ) -> Option<&NmActiveConnection> {
+        self.nm_acs_index
+            .get(&(
+                iface.name.to_string(),
+                NmIfaceType::from(&iface.iface_type),
+            ))
+            .map(Rc::as_ref)
+    }
+
+    pub(crate) fn saved_iter(&self) -> impl Iterator<Item = &NmConnection> {
+        self.nm_saved_cons.values().map(Rc::as_ref)
+    }
+
+    pub(crate) fn is_uuid_activated(&self, uuid: &str) -> bool {
+        self.nm_acs.contains_key(uuid)
+    }
+
+    pub(crate) fn get_applied_by_name_type(
+        &self,
+        name: &str,
+        iface_type: &NmIfaceType,
+    ) -> Option<&NmConnection> {
+        self.nm_applied_conns_index
+            .get(&(name.to_string(), iface_type.clone()))
+            .map(Rc::as_ref)
+    }
+
+    pub(crate) fn get_saved_by_name_type(
+        &self,
+        name: &str,
+        iface_type: &NmIfaceType,
+    ) -> Vec<&NmConnection> {
+        if let Some(nm_conns) = self
+            .nm_saved_conns_index
+            .get(&(name.to_string(), iface_type.clone()))
+        {
+            nm_conns.iter().map(Rc::as_ref).collect()
+        } else {
+            Vec::new()
+        }
+    }
+
+    pub(crate) fn is_activated(&self, uuid: &str) -> bool {
+        self.nm_acs.contains_key(uuid)
+    }
+
+    pub(crate) fn get_applied_by_uuid(
+        &self,
+        uuid: &str,
+    ) -> Option<&NmConnection> {
+        self.nm_applied_cons.get(uuid).map(Rc::as_ref)
+    }
+}
+
+/// Find interface name for NmConnection:
+///  * `connection.interface-name`
+///  * For VLAN, use `vlan.id` and `vlan.parent` if `connection.interface-name`
+///    is empty
+///  * Resolves `802-3-ethernet.mac-address` if defined
+fn get_nm_connection_iface_name(
+    nm_conn: &NmConnection,
+    cur_state: &NetworkState,
+) -> Option<String> {
+    if let Some(iface_name) = nm_conn.iface_name() {
+        if !iface_name.is_empty() {
+            return Some(iface_name.to_string());
+        }
+    }
+
+    if let (Some(vlan_parent), Some(vlan_id)) = (
+        nm_conn.vlan.as_ref().and_then(|v| v.parent.as_deref()),
+        nm_conn.vlan.as_ref().and_then(|v| v.id),
+    ) {
+        return Some(format!("{vlan_parent}.{vlan_id}"));
+    }
+
+    if let Some(mac) = nm_conn
+        .wired
+        .as_ref()
+        .and_then(|s| s.mac_address.as_deref())
+    {
+        if let Some(name) = cur_state.interfaces.iter().find_map(|iface| {
+            if iface.base_iface().mac_address.as_deref() == Some(mac) {
+                Some(iface.name().to_string())
+            } else {
+                None
+            }
+        }) {
+            return Some(name);
+        }
+    }
+    None
```

Comment:
For VLAN, MACVLAN and MACSEC, ethernet.mac-address is the MAC of the parent, not of the interface itself.